core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockInventoryTransactionTranslator.java

```diff
@@ -347,6 +352,46 @@ public void translate(GeyserSession session, InventoryTransactionPacket packet)
 
                         ServerboundUseItemPacket useItemPacket = new ServerboundUseItemPacket(Hand.MAIN_HAND, session.getWorldCache().nextPredictionSequence());
                         session.sendDownstreamPacket(useItemPacket);
+
+                        List<LegacySetItemSlotData> legacySlots = packet.getLegacySlots();
+                        if (packet.getActions().size() == 1 && legacySlots.size() > 0) {
+                            InventoryActionData actionData = packet.getActions().get(0);
+                            LegacySetItemSlotData slotData = legacySlots.get(0);
+                            if (slotData.getContainerId() == 6 && actionData.getToItem().getId() != 0) {
+                                // The player is trying to swap out an armor piece that already has an item in it
+                                // 1.19.4 brings this natively, but we need this specific case for custom head rendering to work
+                                int bedrockHotbarSlot = packet.getHotbarSlot();
+                                Click click = InventoryUtils.getClickForHotbarSwap(bedrockHotbarSlot);
+                                if (click != null && slotData.getSlots().length != 0) {
+                                    Inventory playerInventory = session.getPlayerInventory();
+                                    // Bedrock sends us the index of the slot in the armor container; armor in Java
+                                    // Edition is offset by 5 in the player inventory
+                                    int armorSlot = slotData.getSlots()[0] + 5;
+                                    if (armorSlot == 5) {
+                                        GeyserItemStack armorSlotItem = playerInventory.getItem(armorSlot);
+                                        if (armorSlotItem.getJavaId() == 1055) {
+                                            // checking for player head in the head slot; if there is, we need to manually replace it & apply old skin
+                                            // just calling FakeHeadProvider.restoreOriginalSkin does work too, but causes a visual duplication glitch
+                                            // when trying to re-equip the head. Which should not be possible, but for some reason, is..
+                                            GeyserItemStack hotbarItem = playerInventory.getItem(playerInventory.getOffsetForHotbar(bedrockHotbarSlot));
+
+                                            FakeHeadProvider.restoreOriginalSkin(session, session.getPlayerEntity());
+                                            playerInventory.setItem(armorSlot, hotbarItem, session);
+                                            playerInventory.setItem(bedrockHotbarSlot, armorSlotItem, session);
+
+                                            Int2ObjectMap<ItemStack> changedSlots = new Int2ObjectOpenHashMap<>(2);
```

Comment:
Are these inventory lines still necessary?

Reply:
unfortunately yes; see the discord link i've cc'd